### PR TITLE
Backports 0.18 pr18

### PR DESCRIPTION
**Pull request by Munkybooty**
Still experiencing build issues so again, i do apologize if the pr is bad. There are some parts of the codebase i haven't dealt with before so apologies for any trivial mistakes that normally would not be present.

### CODE REVIEW

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Comment by Munkybooty** [comment]
I've tried to recreate the test failure but i continually have it passing on linux64. I'm not sure whats causing CI to freak out at the moment over `wallet_dump.py`

File: src/wallet/rpcdump.cpp [hunk 1]
```diff
@@ -1332,21 +1333,23 @@ static UniValue ProcessImportDescriptor(ImportData& import_data, std::map<CKeyID
...
+        for (const auto &key_pair: out_keys.pubkeys) {
+            ordered_pubkeys.push_back(key_pair.first);
+        }
 
-    for (const auto& x : out_keys.scripts) {
-        import_data.import_scripts.emplace(x.second);
-    }
+        for (const auto &x: out_keys.scripts) {
+            import_data.import_scripts.emplace(x.second);
+        }
```
Comment (PastaPastaPasta) [inline src/wallet/rpcdump.cpp:1349]: both of these references should be left aligned

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
pls see below + https://github.com/UdjinM6/dash/commits/pr4638

**Comment by UdjinM6** [comment]
I'm just going to repeat: pls see https://github.com/UdjinM6/dash/commits/pr4638 🤷‍♂️

**Comment by Munkybooty** [comment]
Squashed changes to make review easier

**Review by UdjinM6 — APPROVED** [review]
ACK

**Review by UdjinM6 — APPROVED** [review]
re-ACK

File: test/functional/wallet_dump.py [hunk 2]
```diff
@@ -87,12 +86,6 @@ def skip_test_if_missing_module(self):
         self.skip_if_no_wallet()
 
     def setup_network(self):
-        # TODO remove this when usehd=1 becomes the default
-        # use our own cache and -usehd=1 as extra arg as the default cache is run with -usehd=0
-        self.options.tmpdir = os.path.join(self.options.tmpdir, 'hd')
-        self.options.cachedir = os.path.join(self.options.cachedir, 'hd')
-        self._initialize_chain(extra_args=self.extra_args[0])
-        self.set_cache_mocktime()
```
Comment (PastaPastaPasta) [inline test/functional/wallet_dump.py:95]: 15415 why are these changes included

File: src/miner.h [hunk 3]
```diff
@@ -1,17 +1,19 @@
 // Copyright (c) 2009-2010 Satoshi Nakamoto
-// Copyright (c) 2009-2015 The Bitcoin Core developers
+// Copyright (c) 2009-2018 The Bitcoin Core developers
```
Comment (PastaPastaPasta) [inline src/miner.h:2]: should've bumped to 2019

File: test/functional/mining_basic.py [hunk 4]
```diff
@@ -37,9 +40,23 @@ def assert_template(node, block, expect, rehash=True):
 class MiningTest(BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 2
-        self.setup_clean_chain = False
+        self.setup_clean_chain = True
+
+    def mine_chain(self):
+        self.log.info('Create some old blocks')
+        for _ in range(0, 200):
```
Comment (PastaPastaPasta) [inline test/functional/mining_basic.py:47]: not sure why we changed this logic

File: src/wallet/rpcwallet.cpp [hunk 5]
```diff
@@ -2591,9 +2581,7 @@ static UniValue getwalletinfo(const JSONRPCRequest& request)
     obj.pushKV("timefirstkey", pwallet->GetTimeFirstKey());
     obj.pushKV("keypoololdest", pwallet->GetOldestKeyPoolTime());
     obj.pushKV("keypoolsize",   (int64_t)pwallet->KeypoolCountExternalKeys());
-    if (fHDEnabled) {
-        obj.pushKV("keypoolsize_hd_internal",   (int64_t)(pwallet->KeypoolCountInternalKeys()));
-    }
+    obj.pushKV("keypoolsize_hd_internal",   (int64_t)(pwallet->KeypoolCountInternalKeys()));
```
Comment (PastaPastaPasta) [inline src/wallet/rpcwallet.cpp:2584]: why'd we get rid of this check?

Comment (UdjinM6) [reply src/wallet/rpcwallet.cpp:2584]: With 14075 we can use `importmulti` to import pubkeys via `desc` option into _internal_ keypool even for non-HD/empty wallets. Another option would be to can go with `if (pwallet->CanSupportFeature(FEATURE_HD)) {` (like in btc) here and require users to upgrade their wallets for this to work correctly.

File: test/functional/wallet_dump.py (hunk 2 above)
Comment (UdjinM6) [reply test/functional/wallet_dump.py:95]: This was a workaround for the hd/non-hd wallets conflict (cache is created with non-hd). 15415 fixed it, so this hack is no longer need.

File: test/functional/mining_basic.py (hunk 4 above)
Comment (UdjinM6) [reply test/functional/mining_basic.py:47]: Have to bump mocktime on all nodes because otherwise other nodes will ban us for relaying blocks from the future.

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit
